middleware/src/middleware.go

```diff
@@ -21,14 +26,16 @@ type SampleInfo struct {
 type Middleware struct {
 	info        SampleInfo
 	environment system.Environment
-	events      chan interface{}
+	events      chan proto.Message
+	mu          sync.RWMutex
 }
 
 // NewMiddleware returns a new instance of the middleware
 func NewMiddleware(bitcoinRPCUser, bitcoinRPCPassword, bitcoinRPCPort, lightningRPCPath, electrsRPCPort, network string) *Middleware {
 	middleware := &Middleware{
 		environment: system.NewEnvironment(bitcoinRPCUser, bitcoinRPCPassword, bitcoinRPCPort, lightningRPCPath, electrsRPCPort, network),
-		events:      make(chan interface{}),
+		//TODO(TheCharlatan) find a better way to increase the channel size
+		events: make(chan proto.Message, 2), //the channel size needs to be increased every time we had an extra endpoint
```

Comment:
Why do we need a buffered channel here in the first place?
The channel size doesn't correspond to the number of message types/endpoints anyway right, but the number of messages?
So if the base sent multiple messages of the same type, the channel would still block.

Any reason no to use the `WriteMessage` and `ReadMessage` functions that come with the ws library? https://github.com/gorilla/websocket/blob/master/conn.go#L738

Reply:
The reason why the buffered channels were required was that the case when an incoming SystemEnv request was handled, it would potentially emit an event into a blocked channel. This would result in deadlock, since the SystemEnv handling itself would block reading from the event until the write was successful.
SystemEnv now returns a byte array value instead of emitting an event.

Reply:
I see the need for the buffered channel in that case, but I still don't quite get why use a channel in here in the first place. Why not just send the marshaled proto messages/events through the websocket right away with Read and Write?

Reply:
I'm not sure how this could be done in a sensible way without events. At least on the read part, you need something that propagates an event from the websocket read loop. On the write part it is mostly convenience though. It allows you to abstract any networking and related error handling away from the actual logic being executed. One can then separate the flow into actual local base operations (as managed by package middleware), routing information between app and base (as managed by the package handlers) and actually running the websocket with the encryption and error handling in place. 
The events are also a first step towards supporting multiple clients at once with a single middleware backend. To send a notification from the backend, I'd like to emit a single event that then in return is able to trigger events for each of the connected clients that then receive on their send channels and write data to the websocket.